pkg/image/admission/imagepolicy/accept.go

```diff
@@ -0,0 +1,111 @@
+package imagepolicy
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/admission"
+	apierrs "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/util/validation/field"
+
+	"github.com/openshift/origin/pkg/api/meta"
+	"github.com/openshift/origin/pkg/image/admission/imagepolicy/rules"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+var errRejectByPolicy = fmt.Errorf("this image is prohibited by policy")
+
+func accept(accepter rules.Accepter, resolver imageResolver, m meta.ImageReferenceMutator, attr admission.Attributes, covered policyAttributes, excludedRules sets.String) error {
+	gr := attr.GetResource().GroupResource()
+	requiresImage := accepter.RequiresImage(gr)
+	resolvesImage := accepter.ResolvesImage(gr)
+
+	errs := m.Mutate(func(image string) (string, error) {
+		ok, newImage, err := acceptImage(accepter, resolver, image, gr, covered, requiresImage, resolvesImage, excludedRules)
+		if err != nil {
+			return image, err
+		}
+		if !ok {
+			return image, errRejectByPolicy
+		}
+		return newImage, nil
+	})
+
+	for i := range errs {
+		if errs[i].Detail == errRejectByPolicy.Error() {
+			errs[i].Type = field.ErrorTypeForbidden
+		} else {
+			errs[i].Detail = fmt.Sprintf("this image is prohibited by policy: %s", errs[i].Detail)
+		}
+	}
+
+	if len(errs) > 0 {
+		glog.V(5).Infof("failed to create: %v", errs)
+		return apierrs.NewInvalid(attr.GetKind().GroupKind(), attr.GetName(), errs)
+	}
+	glog.V(5).Infof("allowed: %#v", attr)
+	return nil
+}
+
+func acceptImage(accepter rules.Accepter, resolver imageResolver, name string, gr unversioned.GroupResource, covered policyAttributes, requiresImage, resolvesImage bool, excludedRules sets.String) (bool, string, error) {
+	// only resolve the image once per pod
+	if covered.Has(name) {
```

Comment:
yes, will fix that

On Tue, Aug 9, 2016 at 9:09 AM, David Eads <notifications@github.com> wrote:

> In pkg/image/admission/imagepolicy/accept.go
> <https://github.com/openshift/origin/pull/8995#discussion_r74053777>:
>
> > +		} else {
> > +			errs[i].Detail = fmt.Sprintf("this image is prohibited by policy: %s", errs[i].Detail)
> > +		}
> > +	}
> > +
> > +	if len(errs) > 0 {
> > +		glog.V(5).Infof("failed to create: %v", errs)
> > +		return apierrs.NewInvalid(attr.GetKind().GroupKind(), attr.GetName(), errs)
> > +	}
> > +	glog.V(5).Infof("allowed: %#v", attr)
> > +	return nil
> > +}
> > +
> > +func acceptImage(accepter rules.Accepter, resolver imageResolver, name string, gr unversioned.GroupResource, covered policyAttributes, requiresImage, resolvesImage bool, excludedRules sets.String) (bool, string, error) {
> > +	// only resolve the image once per pod
> > +	if covered.Has(name) {
>
> Since you remember the original name, this could match the original name
> and skip pinning it to a digest, couldn't it?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/8995/files/f0c24e731883095660d13951792d6b39faa8018e#r74053777>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p4SRebURr33TR6HS0X_-Fkc7sbQjks5qeHvtgaJpZM4IlDlb>
> .
>